### PR TITLE
Automated cherry pick of #115377: Fetch go version using gimme if needed

### PR DESCRIPTION
**Pull request by liggitt**
Cherry pick of #115377 on release-1.25.

#115377: Fetch go version using gimme if needed

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
The go version defined in `.go-version` is now fetched when invoking test, build, and code generation targets if the current go version does not match it. Set $FORCE_HOST_GO=y while testing or building to skip this behavior, or set $GO_VERSION to override the selected go version.
```

### CODE REVIEW

**Comment by liggitt** [comment]
green go-compatibility unit and integration tests demonstrate release-1.25 builds and passes tests on libraries with the original go version for 1.25 🎉 (not surprising, since 1.25 started on go1.19)

**Comment by liggitt** [comment]
Even before merge, this PR is useful to exercise the -go-compatibility presubmits.

Merging this will simplify testing of new go versions, which https://github.com/kubernetes/enhancements/issues/3744 anticipates happening for 1.25 with an eventual update to go1.20.

That said, I think we should wait to merge this at least until the following occur with https://github.com/kubernetes/kubernetes/pull/115377:
* we go through an actual release on master (e.g. 1.27.0-alpha.2 in a couple weeks) to prove that works with release tooling
* we go through an actual go bump on master (e.g. the go1.20 bump in https://github.com/kubernetes/kubernetes/pull/114502) to prove there's no issues straddling go1.19/go1.20 with this change

/hold

**Comment by liggitt** [comment]
/hold cancel

cc @dims for lgtm

**Comment by liggitt** [comment]
cc @kubernetes/release-managers

**Review by cpanato — APPROVED** [review]
/triage accepted
/priority important-soon
/lgtm

**Pull request merged**
